src/thing_creature.c

```diff
@@ -980,337 +1102,486 @@ void first_apply_spell_effect_to_thing(struct Thing *thing, SpellKind spell_idx,
             if (external_set_thing_state(thing, CrSt_CreatureCombatFlee))
             {
                 cctrl->flee_start_turn = game.play_gameturn;
-                fill_spell_slot(thing, i, spell_idx, duration);
-                cctrl->spell_flags |= spconf->spell_flags;
             }
         }
-    }
-    else
-    if (spconf->spell_flags == CSAfF_Chicken)
-    {
-        if ((get_creature_model_flags(thing) & CMF_NeverChickens) == 0)
+        else // If spell is reapplied reset flee_start_turn and state.
         {
-            if (i != -1)
+            cctrl->flee_start_turn = game.play_gameturn;
+            if (get_creature_state_besides_interruptions(thing) != CrSt_CreatureCombatFlee)
             {
-                fill_spell_slot(thing, i, spell_idx, pwrdynst->strength[spell_lev]);
-                external_set_thing_state(thing, CrSt_CreatureChangeToChicken);
-                cctrl->countdown = duration;
-                cctrl->spell_flags |= spconf->spell_flags;
-                if (spconf->aura_effect != 0)
-                {
-                    cctrl->spell_aura = spconf->aura_effect;
-                    cctrl->spell_aura_duration = spconf->duration;
-                }
+                external_set_thing_state(thing, CrSt_CreatureCombatFlee);
             }
         }
-    } else
-    if (spell_idx == SplK_Light)
+        affected = true;
+    }
+    // TELEPORT.
+    if (flag_is_set(spconf->spell_flags, CSAfF_Teleport)
+    && (!creature_is_immune_to_spell_effect(thing, CSAfF_Teleport)))
+    {
+        if (!creature_under_spell_effect(thing, CSAfF_Teleport))
+        {
+            set_flag(cctrl->spell_flags, CSAfF_Teleport);
+            set_flag(cctrl->stateblock_flags, CCSpl_Teleport);
+        }
+        cctrl->active_teleport_spell = spell_idx; // Remember the spell_idx for a later use.
+        affected = true;
+    }
+    // TIMEBOMB.
+    if (flag_is_set(spconf->spell_flags, CSAfF_Timebomb)
+    && (!creature_is_immune_to_spell_effect(thing, CSAfF_Timebomb)))
+    {
+        if (!creature_under_spell_effect(thing, CSAfF_Timebomb))
+        {
+            set_flag(cctrl->spell_flags, CSAfF_Timebomb);
+            cctrl->timebomb_countdown = duration;
+        }
+        cctrl->active_timebomb_spell = spell_idx; // Remember the spell_idx for a later use.
+        affected = true;
+    }
+    // HEAL.
+    if (flag_is_set(spconf->spell_flags, CSAfF_Heal)
+    && (!creature_is_immune_to_spell_effect(thing, CSAfF_Heal)))
+    {
+        // 'CSAfF_Heal' is only for checking config or immunity, flag is never set to creature.
```

Comment:
Do we want to keep it this way? Should we perhaps give heal a proper duration too?

Reply:
Heal has a duration of 16 by default, let's keep that and spread the healing amount of those [16] turns. And then when it is another value, use that value.

Reply:
I did not consider this behavior change.